File: app/src/main/java/com/iceteaviet/fastfoodfinder/data/DataManager.java
package com.iceteaviet.fastfoodfinder.data;

import android.app.Activity;

import com.iceteaviet.fastfoodfinder.data.domain.store.StoreDataSource;
import com.iceteaviet.fastfoodfinder.data.domain.user.UserDataSource;
import com.iceteaviet.fastfoodfinder.data.prefs.PreferencesHelper;
import com.iceteaviet.fastfoodfinder.data.remote.ClientAuth;
import com.iceteaviet.fastfoodfinder.data.remote.routing.MapsRoutingApiHelper;
import com.iceteaviet.fastfoodfinder.data.remote.store.model.Store;
import com.iceteaviet.fastfoodfinder.data.remote.user.model.User;

import java.util.List;

import io.reactivex.Single;

/**
 * Created by tom on 7/9/18.
 */
public interface DataManager extends ClientAuth {
    StoreDataSource getLocalStoreDataSource();

    StoreDataSource getRemoteStoreDataSource();

    UserDataSource getRemoteUserDataSource();

    UserDataSource getLocalUserDataSource();

    MapsRoutingApiHelper getMapsRoutingApiHelper();

    PreferencesHelper getPreferencesHelper();

    Single<List<Store>> loadStoresFromServer(Activity activity);

    User getCurrentUser();

    void setCurrentUser(User user);
}

File: app/src/main/java/com/iceteaviet/fastfoodfinder/data/DataManager.kt
package com.iceteaviet.fastfoodfinder.data

import com.iceteaviet.fastfoodfinder.data.auth.ClientAuth
import com.iceteaviet.fastfoodfinder.data.domain.prefs.PreferencesRepository
import com.iceteaviet.fastfoodfinder.data.domain.routing.MapsRoutingRepository
import com.iceteaviet.fastfoodfinder.data.domain.store.StoreRepository
import com.iceteaviet.fastfoodfinder.data.domain.user.UserRepository
import com.iceteaviet.fastfoodfinder.data.remote.store.model.Store
import com.iceteaviet.fastfoodfinder.data.remote.user.model.User
import io.reactivex.Single

/**
 * Created by tom on 7/9/18.
 *
 * Central data access point.
 * We may not need this if we use few repository. But it's important when we have many repositories
 */
interface DataManager : ClientAuth, StoreRepository, UserRepository, MapsRoutingRepository, PreferencesRepository {
    fun loadStoresFromServer(): Single<List<Store>>

    fun getCurrentUser(): User?

    fun updateCurrentUser(user: User?)
}
